### PR TITLE
Update transformation to canonical form

### LINKED ISSUE #527: Parsing affine systems in BFFPSV
Note that `BFFPSV` will ignore the `system.c` (even if it accepts this type, so it can be considered a bug). One way is to annotate the type of systems that can be handled by `BFFPSV`, unrelated to this file (i prefer that). 

Another way is to normalize that `CACCS` into a `CLCCS` here in such a way that the input is `U + Singleton(c)` or even `U + c` (a lazy translation).

_Originally posted by @mforets in https://github.com/JuliaReach/Reachability.jl/pull/526_

### LINKED ISSUE #529: Normalize affine ODEs
The methods in `discretize.jl` assume that the passed system is in the "canonical" for `x' = Ax + u`, `u in U`. A system of the form `x' = Ax + Bu` can be brought to this form by mapping the input set `U` into `B*U`.

Now that we have a clean `normalize` step ([after this PR](https://github.com/JuliaReach/Reachability.jl/pull/526),) perhaps it is time to add a function `normalize_ODE` such that it brings the system into the canonical form of above, if the matrix `B` is other than the identity.

Related:

- https://github.com/JuliaReach/Reachability.jl/issues/457

### PR DESCRIPTION
**Pull request by mforets**
Closes #529.
Closes #527.

~~Requires https://github.com/JuliaReach/MathematicalSystems.jl/pull/85.~~ Available in version v0.6.3

### CODE REVIEW

File: src/Utils/normalization.jl [hunk 1]
```diff
@@ -1,68 +1,119 @@
...
+system otherwise.
 
-### Algorithm
+### Notes
 
-We apply the following normalization steps.
+The normalization procedure consists of transforming a given system type into a
+"canonical" format that is used internally. The implementation of `normalize`
+exploits `MathematicalSystems`'s' types, which carry information about the problem
+as a type parameter.
```
Comment (schillic) [inline src/Utils/normalization.jl]: "The implementation ..." is an `### Algorithm` detail.

File: src/Utils/normalization.jl [hunk 2]
```diff
@@ -1,68 +1,119 @@
...
-    return normalize_inputs(system)
-end
+The type union `CF` defines the initial value problems considered canonical,
+i.e. which do not require normalization. More details are given below.
 
-"""
-    normalize_inputs(system)
+Homogeneous ODEs of the form ``x' = Ax`` are canonical if the associated
+problem is a `LinearContinuousSystem` and `A` is a matrix.
+Note that it does not consider constraints on the state-space (such as an
```
Comment (schillic) [inline src/Utils/normalization.jl]:
```suggestion
Note that a `LinearContinuousSystem` does not consider constraints on the state-space (such as an
```

File: src/Utils/normalization.jl [hunk 3]
```diff
@@ -1,68 +1,119 @@
...
+invariant); in this case you may have to use a
```
Comment (schillic) [inline src/Utils/normalization.jl]:
```suggestion
invariant); to specify state constraints, use a
```

File: src/Utils/normalization.jl [hunk 4]
```diff
@@ -1,68 +1,119 @@
...
+The generalization to canonical systems with constraints and possibly time-varying
+non-deterministic inputs is considered next. These systems are of the form
+``x' = Ax + u, u ∈ \\mathcal{U}, x ∈ \\mathcal{X}``. The system type is
+`ConstrainedLinearControlContinuousSystem`, where `A` is a matrix, `X` is a set
+and `U` is an input, that is, any concrete subtype of `AbstractInput`.
 
-### Input
-
-- `system` -- continuous system
+If `U` is not given as an input, normalization accepts either as `LazySet`, or
```
Comment (schillic) [inline src/Utils/normalization.jl]:
```suggestion
If `U` is not given as an input, normalization accepts either a `LazySet`, or
```

File: src/Utils/normalization.jl [hunk 5]
```diff
@@ -1,68 +1,119 @@
...
+a vector of `LazySet`'s. In these cases, the sets are wrapped around an appropriate
```
Comment (schillic) [inline src/Utils/normalization.jl]:
```suggestion
a vector of `LazySet`s. In these cases, the sets are wrapped around an appropriate
```

File: src/Utils/normalization.jl [hunk 6]
```diff
@@ -1,68 +1,119 @@
...
+    X = _wrap_invariant(system.X, n)
+    U = _wrap_inputs(system.U, system.B, system.c)
+    CACDS(system.A, I(n, N), X, U)
 end
 
+_wrap_invariant(X::LazySet, n::Int) = X
+_wrap_invariant(X::Nothing, n::Int) = Universe(n)
+
+_wrap_inputs(U::AbstractInput, B::IdentityMultiple) = U
+_wrap_inputs(U::AbstractInput, B::AbstractMatrix) = map(u -> B*u, U)
```
Comment (schillic) [inline src/Utils/normalization.jl]: What is the result type of `map`?

File: src/solve.jl [hunk 7]
```diff
@@ -56,25 +56,26 @@ end
 solve(system::AbstractSystem, options::Pair{Symbol,<:Any}...) =
     solve(system, Options(Dict{Symbol,Any}(options)))
 
-function solve!(system::InitialValueProblem{<:Union{AbstractContinuousSystem,
-                                                     AbstractDiscreteSystem}},
+function solve!(problem::InitialValueProblem,
```
Comment (schillic) [inline src/solve.jl]: Why remove this?

**Comment by schillic** [comment]
I think a good part of this new documentation should be put somewhere more prominently. Nobody will/should have to read the `normalize` docs. This is rather a new section in the manual.

File: src/Utils/normalization.jl (hunk 2 above)
Comment (mforets) [reply src/Utils/normalization.jl]: Thanks. Actually there is the [ConstrainedLinearContinuousSystem](https://juliareach.github.io/MathematicalSystems.jl/latest/lib/types/#MathematicalSystems.ConstrainedLinearContinuousSystem), so we should support this as well. But i would rather program this in another PR.

Comment (schillic) [reply src/Utils/normalization.jl]: I just wanted to clarify the "it," that's all.

File: src/Utils/normalization.jl (hunk 6 above)
Comment (mforets) [reply src/Utils/normalization.jl]: The result type of `map` is the same as that of `U`, see [`map(f::Function, c::ConstantInput)`](https://github.com/JuliaReach/MathematicalSystems.jl/blob/master/src/inputs.jl#L77) and [`map(f::Function, v::VaryingInput)`](https://github.com/JuliaReach/MathematicalSystems.jl/blob/master/src/inputs.jl#L199).

File: src/solve.jl (hunk 7 above)
Comment (mforets) [reply src/solve.jl]: I thought about keeping the `solve!` interface generic (it's already the case for the other `solve` methods).

**Comment by mforets** [comment]
> This is rather a new section in the manual.

Makes sense. I don't plan to work on the Reachability docs right now so i'd prefer to defer this. For the time being i have added `normalize` to the docs in the last commit.

**Pull request merged**